### PR TITLE
Refactor `check access?` methods for projects and packages

### PR DESCRIPTION
**Pull request by eduardoj**

### CODE REVIEW

File: src/api/app/models/owner_search/assignee.rb [hunk 1]
```diff
@@ -37,7 +37,7 @@ def extract_maintainer_project_level(owner, package)
     end
 
     def extract_maintainer(package)
-      return unless package && Package.check_access?(package)
+      return unless package && package.check_access?
```
[the remark this answers is not in the document]
Comment (eduardoj) [reply src/api/app/models/owner_search/assignee.rb]: Applied.

File: src/api/app/models/concerns/project_maintenance.rb [hunk 2]
```diff
@@ -22,7 +22,7 @@ def get_maintenance_project(at = nil)
       at ||= AttribType.find_by_namespace_and_name!('OBS', 'MaintenanceProject')
       maintenance_project = Project.find_by_attribute_type(at).first
 
-      return unless maintenance_project && check_access?(maintenance_project)
+      return unless maintenance_project && maintenance_project.check_access?
```
Comment (hennevogel) [inline src/api/app/models/concerns/project_maintenance.rb:25]:
```suggestion
      return unless maintenance_project&.check_access?
```

File: src/api/app/models/package.rb [hunk 3]
```diff
@@ -20,6 +20,8 @@ class Package < ApplicationRecord
   accepts_nested_attributes_for :kiwi_image
 
   belongs_to :project, inverse_of: :packages
+
+  delegate :check_access?, to: :project
```
Comment (hennevogel) [inline src/api/app/models/package.rb:24]: Can you rather change the callers? One less layer of indirection...

File: src/api/app/models/concerns/project_maintenance.rb (hunk 2 above)
Comment (eduardoj) [reply src/api/app/models/concerns/project_maintenance.rb:25]: :+1: Applied.

File: src/api/app/models/package.rb (hunk 3 above)
Comment (eduardoj) [reply src/api/app/models/package.rb]: Changed. I added this change in another commit.

File: src/api/app/models/package.rb [hunk 4]
```diff
@@ -117,13 +117,6 @@ class Package < ApplicationRecord
   has_one :backend_package, foreign_key: :package_id, dependent: :destroy, inverse_of: :package # rubocop:disable Rails/RedundantForeignKey
   has_many :tokens, dependent: :destroy
 
-  def self.check_access?(package)
-    return false if package.nil?
-    return false unless package.instance_of?(Package)
```
Comment (saraycp) [inline src/api/app/models/package.rb:120]: I'm not sure if we can get rid of this condition. What happens if you try to do `package.project.check_access` on a `package` that returns something different from `Package`?

Package associated to a LinkedProject or similar:   `[https://github.com/openSUSE/open-build-service/pull/16768/commits/ae65d98712451d5849f89ddcad35d8d4c249bbb7#diff-9c1e6becaa519550173f515e411731cfd1d9810cdd4671c064627421445e0a76L122`](https://github.com/openSUSE/open-build-service/wiki/Links#instantiating-returns-nil-1)

Comment (eduardoj) [reply src/api/app/models/package.rb:120]: True, good catch. `package` could be a `String`. AFAIK, this only happens in one place, inside the `self.get_by_project_and_name` method.

I modified that method accordingly. Please take a look again.

**Review by saraycp — APPROVED** [review]

**Pull request merged**